File: src/main/java/com/openlattice/chronicle/controllers/legacy/ChronicleController.java
package com.openlattice.chronicle.controllers.legacy;

import com.codahale.metrics.annotation.Timed;
import com.google.common.collect.SetMultimap;
import com.openlattice.chronicle.ChronicleApi;
import com.openlattice.chronicle.services.edm.EdmCacheManager;
import com.openlattice.chronicle.services.upload.AppDataUploadManager;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import javax.inject.Inject;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

@RestController
@RequestMapping( ChronicleApi.CONTROLLER )
public class ChronicleController implements ChronicleApi {

    @Inject
    private AppDataUploadManager dataUploadManager;

    @Inject
    private EdmCacheManager edmCacheManager;

    @Override
    @Timed
    @RequestMapping(
            path = STUDY_ID_PATH + PARTICIPANT_ID_PATH + DATASOURCE_ID_PATH,
            method = RequestMethod.POST,
            consumes = MediaType.APPLICATION_JSON_VALUE
    )
    public Integer upload(
            @PathVariable( STUDY_ID ) UUID studyId,
            @PathVariable( PARTICIPANT_ID ) String participantId,
            @PathVariable( DATASOURCE_ID ) String datasourceId,
            @RequestBody List<SetMultimap<UUID, Object>> data ) {

        return dataUploadManager.upload( null, studyId, participantId, datasourceId, data );
    }

    @Override
    @Timed
    @RequestMapping(
            path = EDM_PATH,
            method = RequestMethod.POST,
            consumes = MediaType.APPLICATION_JSON_VALUE,
            produces = MediaType.APPLICATION_JSON_VALUE
    )
    public Map<String, UUID> getPropertyTypeIds( @RequestBody Set<String> propertyTypeFqns ) {
        return edmCacheManager.getHistoricalPropertyTypeIds( propertyTypeFqns );
    }

    @Override
    @Timed
    @RequestMapping(
            path = STATUS_PATH,
            method = RequestMethod.GET
    )
    public Boolean isRunning() {
        //TODO: Ensure connectivity with OpenLattice backend.
        return true;
    }
}

File: src/main/kotlin/com/openlattice/chronicle/controllers/legacy/ChronicleController.kt
package com.openlattice.chronicle.controllers.legacy

import com.codahale.metrics.annotation.Timed
import com.google.common.collect.SetMultimap
import com.openlattice.chronicle.ChronicleApi
import com.openlattice.chronicle.services.enrollment.EnrollmentManager
import com.openlattice.chronicle.services.legacy.LegacyEdmResolver
import com.openlattice.chronicle.services.upload.AppDataUploadManager
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import java.util.*
import javax.inject.Inject

/**
 *
 * @author Matthew Tamayo-Rios &lt;matthew@openlattice.com&gt;
 */
@RestController
@RequestMapping(ChronicleApi.CONTROLLER)
class ChronicleController : ChronicleApi {
    @Inject
    private lateinit var dataUploadManager: AppDataUploadManager

    @Inject
    private lateinit var enrollmentManager: EnrollmentManager

    @Timed
    @RequestMapping(
            path = [ChronicleApi.STUDY_ID_PATH + ChronicleApi.PARTICIPANT_ID_PATH + ChronicleApi.DATASOURCE_ID_PATH],
            method = [RequestMethod.POST], consumes = [MediaType.APPLICATION_JSON_VALUE]
    )
    override fun upload(
            @PathVariable(ChronicleApi.STUDY_ID) studyId: UUID,
            @PathVariable(ChronicleApi.PARTICIPANT_ID) participantId: String,
            @PathVariable(ChronicleApi.DATASOURCE_ID) datasourceId: String,
            @RequestBody data: List<SetMultimap<UUID, Any>>
    ): Int {
        val organizationId = enrollmentManager.getOrganizationIdForLegacyStudy( studyId )
        return dataUploadManager.upload(organizationId, studyId, participantId, datasourceId, data)
    }

    @Timed
    @RequestMapping(
            path = [ChronicleApi.EDM_PATH], method = [RequestMethod.POST],
            consumes = [MediaType.APPLICATION_JSON_VALUE], produces = [MediaType.APPLICATION_JSON_VALUE]
    )
    override fun getPropertyTypeIds(@RequestBody propertyTypeFqns: Set<String>): Map<String, UUID> {
        return LegacyEdmResolver.getLegacyPropertyTypeIds(propertyTypeFqns)
    }

    @Timed
    @RequestMapping(path = [ChronicleApi.STATUS_PATH], method = [RequestMethod.GET])
    override fun isRunning(): Boolean {
        return true
    }
}
